### PR TITLE
wrapnils: add `??.` which returns an `Option`

### PR DESCRIPTION
**Pull request by timotheecour**
* fixes the remaining TODO in https://github.com/nim-lang/Nim/pull/13023 as well as https://github.com/nim-lang/Nim/pull/13023#issuecomment-685337788 by adding `??.` which returns an `Option`

* adds tests (and also we now test in VM too)

### CODE REVIEW

File: lib/std/wrapnils.nim [hunk 1]
```diff
@@ -26,14 +26,19 @@ type Wrapnil[T] = object
...
   ## See top-level example.
   Wrapnil[T](valueImpl: a, validImpl: true)
 
-template unwrap(a: Wrapnil): untyped =
-  ## See top-level example.
+template get*[T](a: Wrapnil[T]): T =
+  ## Unbox `a`.
   a.valueImpl
 
+proc isSome*(a: Wrapnil): bool {.inline.} =
```
Comment (juancarlospaco) [inline lib/std/wrapnils.nim]: Should it be `self` according with the updated nep ?.
Can be `func` ?.

File: lib/std/wrapnils.nim [hunk 2]
```diff
@@ -26,14 +26,19 @@ type Wrapnil[T] = object
...
 
-proc wrapnil[T](a: T): Wrapnil[T] =
+template wrapnil[T](a: T): Wrapnil[T] =
   ## See top-level example.
   Wrapnil[T](valueImpl: a, validImpl: true)
 
-template unwrap(a: Wrapnil): untyped =
-  ## See top-level example.
+template get*[T](a: Wrapnil[T]): T =
+  ## Unbox `a`.
```
Comment (juancarlospaco) [inline lib/std/wrapnils.nim]: Unwrap instead of Unbox ?.
Use unwrap everywhere or unbox everywhere, but unwrap is more common.

**Comment by timotheecour** [comment]
PTAL

**Comment by timotheecour** [comment]
marking as draft after suggestion by @Clyybber to re-use Options type, and also because `get` should probably raise if not isSome (plus maybe have an overload with a default); but that would be all simplified if it just imports Option

File: lib/std/wrapnils.nim [hunk 3]
```diff
@@ -31,13 +31,18 @@ type Wrapnil[T] = object
...
-template unwrap(a: Wrapnil): untyped =
-  ## See top-level example.
-  a.valueImpl
+template get*[T](self: Wrapnil[T]): T =
+  ## Unwraps the value contained in `self`.
+  self.valueImpl
+
+func isSome*(self: Wrapnil): bool {.inline.} =
+  ## Returns true if `self` didn't contain intermediate `nil` values (note that
+  ## `self.get` itself can be nil even in that case)
```
Comment (konsumlamm) [inline lib/std/wrapnils.nim]:
```suggestion
  ## `self.get` itself can be nil even in that case).
```

Comment (timotheecour) [reply lib/std/wrapnils.nim]: done

**Comment by timotheecour** [comment]
@Clyybber PTAL, I've now implemented the variation we talked about (reusing std/options)

File: lib/std/wrapnils.nim [hunk 4]
```diff
@@ -23,89 +23,98 @@ runnableExamples:
...
-template fakeDot*(a: Wrapnil, b): untyped =
+template fakeDot*(a: Option, b): untyped =
   ## See top-level example.
   let a1 = a # to avoid double evaluations
-  let a2 = a1.valueImpl
-  type T = Wrapnil[typeof(a2.b)]
-  if a1.validImpl:
+  type T = Option[typeof(a1.get.b)]
+  if isSome(a1):
+    let a2 = a1.get
```
Comment (ringabout) [inline lib/std/wrapnils.nim]: consider using unsafeGet?

Comment (timotheecour) [reply lib/std/wrapnils.nim]: I don't see how that could work (please show if you have something concrete), you'd loose information on whether there was an intermediate nil in the chain, eg this must hold:
```nim
import std/wrapnils
type Foo = ref object
  x: int
var f: Foo
doAssert not (??.f.x).isSome
doAssert (??.Foo(x: 0).x).isSome
```

Comment (ringabout) [reply lib/std/wrapnils.nim]: ?
```nim
  if isSome(a1):
    let a2 = a1.unsafeGet
```

**Review by Clyybber — APPROVED** [review]

File: lib/std/wrapnils.nim (hunk 4 above)
Comment (timotheecour) [reply lib/std/wrapnils.nim]: I had misunderstood your point, you're correct! I've fixed it here and elsewhere

File: lib/std/wrapnils.nim [hunk 5]
```diff
@@ -23,89 +23,88 @@ runnableExamples:
...
   elif n.len == 0:
-    result = newCall(bindSym"wrapnil", n)
+    result = newCall(bindSym"option", n)
   else:
     n[0] = replace(n[0])
     result = n
 
-macro `?.`*(a: untyped): untyped =
+proc safeGet[T](a: Option[T]): T {.inline.} =
+  get(a, default(T))
```
Comment (Clyybber) [inline lib/std/wrapnils.nim:81]: Nit: Only used once, so might as well inline it.

Comment (timotheecour) [reply lib/std/wrapnils.nim:81]: inlining it as
```
get(`result`, default(typeof(unsafeGet(`result`))))
```

would semcheck `result` twice, which is worse.

- [ ] In future work I'd like to add a new template `options.getAsIs` which would return `val` as-is:
```nim
template getAsIs*(a: Option): auto = a.val
```
then I'd simply use 
```nim
getAsIs(`result`)
```

which is more efficient, less roundabout

**Review by ringabout — APPROVED** [review]

**Pull request merged**